packages/firestore/src/api/database.ts

```diff
@@ -552,6 +553,61 @@ export class WriteBatch
   }
 }
 
+class UntypedFirestoreDataConverterAdapter<U>
+  extends Compat<PublicFirestoreDataConverter<U>>
+  implements UntypedFirestoreDataConverter<U> {
+  private static readonly INSTANCES = new WeakMap();
+
+  private constructor(
+    private readonly _firestore: Firestore,
+    delegate: PublicFirestoreDataConverter<U>
+  ) {
+    super(delegate);
+  }
+
+  fromFirestore(snapshot: unknown, options?: unknown): U {
+    return this._delegate.fromFirestore(
+      new QueryDocumentSnapshot<U>(
+        this._firestore,
+        snapshot as ExpQueryDocumentSnapshot<U>
```

Comment:
This unfortunately causes any data (such as DocumentReferences) to be returned as exp-types. You need to follow the pattern here: https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/api/database.ts#L417